File: Trace2JSON/Trace2JSON/Parsers/ActivityMonitorParser.m
//
//  ActivityMonitorParser.m
//  Trace2JSON
//
//  Created by Itay Brenner on 6/20/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

#import "ActivityMonitorParser.h"
#import "InstrumentsPrivateHeaders.h"

@implementation ActivityMonitorParser

- (NSString *) instrumentName {
    return @"Activity Monitor";
}

- (void)readRow:(XRAnalysisCoreReadCursor *)cursor formatter:(XREngineeringTypeFormatter *)formatter result:(NSMutableDictionary *)result {
    while (XRAnalysisCoreReadCursorNext(cursor)) {
        BOOL objectFound = NO;
        XRAnalysisCoreValue *object = nil;
        objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 0, &object);
        NSString *time = objectFound ? [formatter stringForObjectValue:object] : @"";
        objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 2, &object);
        NSString *process = objectFound ? [formatter stringForObjectValue:object] : @"";
        objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 6, &object);
        NSString *cpu = objectFound ? [formatter stringForObjectValue:object] : @"";
        objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 10, &object);
        NSString *memory = objectFound ? [formatter stringForObjectValue:object] : @"";
        objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 8, &object);
        NSString *threads = objectFound ? [formatter stringForObjectValue:object] : @"";
        
        NSMutableDictionary *processDictionary = nil;
        if([result objectForKey:process]) {
            processDictionary = [result objectForKey:process];
        } else {
            processDictionary = [NSMutableDictionary new];
        }
        
        NSDictionary *rowDict = @{
            @"cpu": cpu,
            @"threads": threads,
            @"memory": memory
        };
        [processDictionary setObject:rowDict forKey:time];
        
        [result setObject:processDictionary forKey:process];
    }
}

- (id) parseContext:(NSArray<XRContext *> *) contexts withRun:(XRRun *)run {
    NSMutableDictionary *result = [NSMutableDictionary new];
    
    XRContext *context = contexts[0];
    [context display];
    XRAnalysisCoreTableViewController *controller = Ivar(context.container, _tabularViewController);
    XRTime duration = run.timeRange.length;
    for (XRTime time = 0; time < duration; time += NSEC_PER_SEC) {
        [controller setDocumentInspectionTime:time];
        [controller _retrieveResponse];
        XRAnalysisCorePivotArray *array = controller._currentResponse.content.rows;
        if (array) {
            XREngineeringTypeFormatter *formatter = IvarCast(array.source, _filter, XRAnalysisCoreTableQuery * const).fullTextSearchSpec.formatter;
            [array access:^(XRAnalysisCorePivotArrayAccessor *accessor) {
                [accessor readRowsStartingAt:0 dimension:0 block:^(XRAnalysisCoreReadCursor *cursor) {
                    [self readRow:cursor formatter:formatter result:result];
                }];
            }];
        }
    }
    
    
    
    return result;
}

@end

File: Trace2JSON/Trace2JSON/Parsers/ActivityMonitorParser.swift
//
//  ActivityMonitorParser.swift
//  Trace2JSON
//
//  Created by Itay Brenner on 6/22/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

import Foundation

class ActivityMonitorParser: ParserProtocol {
    func parseContext(_ contexts: [XRContext], run: XRRun) -> InstrumentRun {
        let activityRun = ActivityRun()
        activityRun.run = Int(run.runNumber())
        activityRun.runName = run.displayName()
        
        let context = contexts[0]
        context.display()
        
        guard let container = context.container() as? XRAnalysisCoreDetailViewController,
            let controller: XRAnalysisCoreTableViewController = RuntimeHacks.getIvar(instance: container, name: "_tabularViewController") else {
                return activityRun
        }
        
        let duration = run.timeRange().length
        for time in stride(from: 0, to: duration, by: UInt64.Stride(NSEC_PER_SEC)) {
            controller.setDocumentInspectionTime(time)
            controller._retrieveResponse()
            
            guard let content = controller._currentResponse().content,
                let array = content.rows(),
                let filter: XRAnalysisCoreTableQuery = RuntimeHacks.getIvar(instance: array.source()!, name: "_filter"),
                let formatter = filter.fullTextSearchSpec()!.formatter() else {
                    continue
            }
            
            array.access({ (accessor) in
                accessor?.readRowsStarting(at: 0, dimension: 0, block: { (cursor) in
                    self.readRow(cursor, formatter, activityRun)
                })
            })
        }
        
        return activityRun
    }
    
    fileprivate func readRow(_ cursor: UnsafeMutableRawPointer?, _ formatter: XREngineeringTypeFormatter, _ activityRun: ActivityRun) {
        while(XRAnalysisCoreReadCursorNext(cursor)) {
            var objectFound: Int32 = 0
            var object: XRAnalysisCoreValue? = nil
            
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 0, &object)
            let time = objectFound != 0 ? formatter.string(for: object) : ""
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 2, &object)
            let process = objectFound != 0 ? formatter.string(for: object) : ""
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 6, &object)
            let cpu = objectFound != 0 ? formatter.string(for: object) : ""
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 10, &object)
            let memory = objectFound != 0 ? formatter.string(for: object) : ""
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 8, &object)
            let threads = objectFound != 0 ? formatter.string(for: object) : ""
            
            let row = ActivityRow(time: time ?? "",
                               process: process ?? "",
                               cpu: cpu ?? "",
                               memory: memory ?? "",
                               threads: threads ?? "")
            
            if let processActivity = activityRun.result[process!] {
                processActivity.activity[time!] = row
            } else {
                let processActivity = ProcessActivity()
                processActivity.activity[time!] = row
                
                activityRun.result[process!] = processActivity
            }
        }
    }
}
